omnipaxos_storage/src/persistent_storage.rs

```diff
@@ -0,0 +1,587 @@
+            s: PhantomData::default(),
+        }
+    }
+
+    // Creates a new storage instance, panics if a commitlog or rocksDB/sled instance exists in the given path
+    pub fn new(storage_config: PersistentStorageConfig) -> Self {
+        let path = storage_config
+            .path
+            .as_ref()
+            .expect("No path found in config");
+
+        std::fs::metadata(format!("{path}{COMMITLOG}")).expect_err(&format!(
+            "Cannot create new instance, commitlog already exists in {}",
+            path
+        ));
+        std::fs::metadata(format!("{path}{DATABASE}")).expect_err(&format!(
+            "Cannot create new instance, database already exists in {}",
+            path
+        ));
+
+        Self::open(storage_config)
+    }
+}
+
+impl<T, S> Storage<T, S> for PersistentStorage<T, S>
+where
+    T: Entry + Serialize + for<'a> Deserialize<'a>,
+    S: Snapshot<T> + Serialize + for<'a> Deserialize<'a>,
+{
+    fn append_entry(&mut self, entry: T) -> u64 {
+        let entry_bytes = bincode::serialize(&entry).expect("Failed to serialize log entry");
+        let offset = self
+            .commitlog
+            .append_msg(entry_bytes)
+            .expect("Failed to append log entry");
+        self.commitlog.flush().expect("Failed to flush Commitlog"); // ensure durable writes
+        offset + 1 // +1 as commitlog returns the offset the entry was appended at, while we should return the index that the entry got in the log.
+    }
+
+    fn append_entries(&mut self, entries: Vec<T>) -> u64 {
+        let serialized = entries
+            .into_iter()
+            .map(|entry| bincode::serialize(&entry).expect("Failed to serialize log entries"));
+        let offset = self
+            .commitlog
+            .append(&mut MessageBuf::from_iter(serialized))
+            .expect("Falied to append log entries");
+        self.commitlog.flush().expect("Failed to flush Commitlog"); // ensure durable writes
+        offset.first() + offset.len() as u64
+    }
+
+    fn append_on_prefix(&mut self, from_idx: u64, entries: Vec<T>) -> u64 {
+        self.commitlog
+            .truncate(from_idx)
+            .expect("Failed to truncate log");
+        self.append_entries(entries)
+    }
+
+    fn get_entries(&self, from: u64, to: u64) -> Vec<T> {
+        let buffer = self
+            .commitlog
+            .read(from, ReadLimit::default())
+            .expect("Failed to read from replicated log");
+        let mut entries = Vec::<T>::with_capacity((to - from) as usize);
+        let mut iter = buffer.iter();
+        for _ in from..to {
+            match iter.next() {
+                Some(msg) => {
+                    entries.push(
+                        bincode::deserialize(msg.payload())
+                            .expect("Failed to deserialize log entries"),
+                    );
+                }
+                None => {
+                    return vec![]; // early return here
```

Comment:
Is this early return here correct? Seems like it is filling things up into ``entries`` then once hitting None on the iterator, it returns an empty Vec instead of ``entries``

Reply:
The function is correct according to the current API. The `get_entries()` is currently defined such that if the user tries to read from an index out of bounds, an empty vector is returned. This might be a bit counterintuitive so it might change in the future.

@AbyelT However, would it be possible first to check if all entries of the requested range really exist in the log before starting to read? So we don't start reading in the commitlog and then anyway end up returning `vec![]`.

Reply:
Sure, the function can check if the endpoint `to` is greater than the current log length and thus out of range. That way we can move the early return to the beginning in the function.